### PR TITLE
Fix waving water not rendered correctly through glass underwater 

### LINKED ISSUE #3733: Waving water and new style water bugs
Waving water (new style water) has some side-effects, like this https://i.imgur.com/OpAGUyj.jpg
I want to point out another problem with it, it is observed underwater from a building/tunnel when looking through a glass, you will see that water renders only partially as in screenshot below (it is like volume indicator, filled only partially, 70-80% or so and doing wave motion). Disabling "waving water" solves the issue.

http://i.imgur.com/eH4zNHo.png

### PR DESCRIPTION
**Pull request by sfence**
- Goal of the PR
Fixes #3733.
- How does the PR work?
Liquid node sides in contact with the glass are not rendered if the camera is in the same liquid type.
- Does it resolve any reported issue?
Fix bug #3733.

## To do

This PR is a Ready for Review.

## How to test

Go to the world, find water, and place glass under water.

https://github.com/minetest/minetest/assets/17455197/7fb18bcf-25bc-44a8-9528-3579e39a196f

### CODE REVIEW

**Comment by appgurueu** [comment]
Thanks for the bugfix. 

I now see what you mean with a "full scene refresh": Regenerating all mapblock meshes to show/hide some faces depending on which liquid the camera is in. I agree that this is somewhat problematic.

If we decide to keep this approach, we can still avoid regenerating mapblock meshes. The best way to go about this probably is to give these faces separate materials. This would let us show / hide them by either binding the appropriate texture or a blank texture; it should work fine with the "old" (fixed, shader-less) rendering pipeline. (On the "modern" rendering pipeline, we might be able to use a geometry shader to suppress these faces to begin with by setting an appropriate uniform. But I don't think such an optimization is necessary, the blank texture should be fine.)

That being said, I would prefer it if we could figure out a different criterion for rendering these faces (though I'm not yet sure what that would be, or if it's even possible). The way we implement water is basically a hack. What we *want* to do is render just the "exterior" surfaces; if a water source is enclosed by glass, we should render all its faces; if on the other hand glass is enclosed by water sources, we do not want to add water faces there.

**Comment by sfence** [comment]
@appgurueu 
Doing the check via camera position is probably the most effective way to determine if that should or should not be shown.
The problem is that the glass area can be big, can have air in it, and can be particularly above water...

**Comment by sfence** [comment]
@appgurueu
Now only generated MapBlockMeshes is modified. Work better.

**Comment by sfence** [comment]
I can probably integrate a fix for #3075 into this if there is a consensus about a solution for it. It is related to the same part of code in the `drawSolidNode` method.

**Comment by appgurueu** [comment]
> I can probably integrate a fix for #3075 into this if there is a consensus about a solution for it. It is related to the same part of code in the `drawSolidNode` method.

I would prefer it if this PR stays focused on this issue for now.

(Regarding "Is it acceptable to have waving water on the roof of sunken caves also?", my answer would be "no".)

Sorry for the long wait times, I'll try to review this properly in ~2 weeks.

**Review by appgurueu — COMMENTED** [review]
### Expected Behavior

I think unfortunately it's a bit hard to put the finger on what the general rule for the expected behavior should be, mostly due to the fact that the way we implement water is a bit of a hack (albeit a hack not uncommon in computer graphics). In reality, water visuals are a mix of refraction and reflection, whereas in Minetest's rendering, we "approximate" this by virtue of a cheap semitransparent water "surface" + blending (and the post effect color).

It seems to me that we want to render "exterior" water "surface" faces while not rendering "interior" faces. Unfortunately, determining what is "interior" and what is "exterior" can't be done locally (if at all); you can't distinguish a water pocket enclosed by glass vs. a glass pocket enclosed by water locally.

If we're within water, and our line of sight passes through glass blocks within water, we expect there to be no water faces next to the glass blocks, otherwise the blue-ish tint of water would be applied multiple times.

If we're however outside of water, we expect the faces to be rendered; there's no post effect color on our screen.

And then there are of course the cases where our line of sight goes water - glass - air - glass - water or similar. In such cases, it doesn't look right to hide the glass - water faces while showing air - water faces. (See the water pocket screenshot).

I think it's probably fine to use an approximation that works well in most of the cases. From my understanding, this PR works by hiding the faces of the liquid next to glass the player's camera is currently in, which doesn't seem unreasonable - you don't want to apply the "water tint" twice, once in the form of the post effect color, and once in the form of the water texture.

### Testing

At first, I tested with waving water disabled.

#### Glass "Border"

I expect the "exterior" water faces to be visible here, as in this image:

Screenshot

With this PR, there is an "edge case" where they will be incorrectly hidden if I look at the water from inside a water pocket. I'm not sure whether this regression is acceptable to me.

Screenshot

This is inherent to the logic in this PR and looks the same with waving water.

#### Glass "Tower"

The tower looks about as right as it can look for me

Screenshot

In particular, it looks fine from underwater:

Screenshot from 2024-03-02 19-45-17

even with waving water enabled, which is precisely the issue this PR aims to fix.

#### Glass "Air Pocket"

Because the camera is inside air, the water faces surrounding the pocket are rendered, as they should be. This is with waving water enabled.

Screenshot from 2024-03-02 22-23-56

### Code

The code looks okay except for some nitpicks (see my review comments). I rebased it; you can find my rebased branch [here](https://github.com/appgurueu/minetest/tree/sfence_fix_3733).

### TL;DR

I'm not sure what to do with this fix. It seems to address the linked issue, but in exchange for another one (and a relatively tiny bit of added complexity), which isn't necessarily a good trade. I wonder whether there's a better solution that's also feasible to implement. It would be good to get some feedback here; what do other games do, what do players expect to see?

File: src/client/clientmap.cpp [hunk 1]
```diff
@@ -156,6 +158,30 @@ void ClientMap::updateCamera(v3f pos, v3f dir, f32 fov, v3s16 offset, video::SCo
 	// reorder transparent meshes when camera crosses node boundary
 	if (previous_node != current_node)
 		m_needs_update_transparent_meshes = true;
+
+	// regenerate mapblocks meshes if neccessary
```
[the remark this answers is not in the document]
Comment (sfence) [reply src/client/clientmap.cpp]: Comment updated.

File: src/client/content_mapblock.cpp [hunk 2]
```diff
@@ -425,6 +426,7 @@ void MapblockMeshGenerator::drawSolidNode()
 		MapNode neighbor = data->m_vmanip.getNodeNoEx(p2);
 		content_t n2 = neighbor.getContent();
 		bool backface_culling = cur_node.f->drawtype == NDT_NORMAL;
+		bool is_hiddable = false;
```
[the remark this answers is not in the document]
Comment (sfence) [reply src/client/content_mapblock.cpp]: Fixed.

File: src/client/clientmap.cpp [hunk 3]
```diff
@@ -399,6 +425,7 @@ void ClientMap::updateDrawList()
 				} else if (mesh) {
 					// without mesh chunking we can add the block to the drawlist
```
[the remark this answers is not in the document]
Comment (sfence) [reply src/client/clientmap.cpp:436]: moved to the private method.

File: src/client/mapblock_mesh.cpp [hunk 4]
```diff
@@ -917,6 +932,38 @@ bool MapBlockMesh::animate(bool faraway, float time, int crack,
...
+{
+	// Texture animation
+	for (auto &it : m_hiddable_info) {
+		const TileLayer &tile = it.second.tile;
+
+		scene::IMeshBuffer *buf = m_mesh[it.first.first]->getMeshBuffer(it.first.second);
+
+		bool hide = (it.second.liquid_source_id == liquid_source_id) &&
+				(it.second.liquid_flowing_id == liquid_flowing_id);
+		if (it.second.hidden != hide) {
```
[the remark this answers is not in the document]
Comment (sfence) [reply src/client/mapblock_mesh.cpp:949]: Refactored.

**Comment by Zughy** [comment]
@sfence rebase needed

**Comment by rubenwardy** [comment]
Looks like this is quite outdated by this point. Rebase needed if this is still relevant
